### DISCUSSION TITLE
Official HyperHDR add-on for LibreELEC

### DISCUSSION THREAD

**Discussion by awawa-dev** [awawa-dev/HyperHDR#600] [Show and tell]
To celebrate the return of Raspberry Pi 4 availability on the market, I have prepared an official native LibreELEC HyperHDR v20beta addon for RPi4. Copy it on pendrive and just install via Kodi Addons (Kodi: System ⇒ Add-ons ⇒ install from zip file). It works right after installing, no need to manually modify system files/services. Note that you still need a USB grabber to capture video. LED drivers such as HyperSerialPico work fine.

As a bonus add-on was also test compiled for PC platform: generic PC x86_64 (PROJECT=Generic DEVICE=Generic ARCH=x86_64). In return, I am asking for feedback because I do not use this configuration. 

> [!IMPORTANT]
> Latest plugins are using HyperHDR v20. Available for LibreELEC 11 & 12.

----

# Download

Download ready to use add-ons for Raspberry Pi 4 and standard PC (x86_64) LibreELEC 11 & 12 platforms: https://github.com/awawa-dev/LibreELEC.tv/releases

You can find the add-on sources here on my LibreELEC 11 fork: https://github.com/awawa-dev/LibreELEC.tv/tree/libreelec-11.0-hyperhdr
and LibreELEC 12 fork: https://github.com/awawa-dev/LibreELEC.tv/tree/libreelec-12.0-hyperhdr

----

# Compiling

If you need it for another platform, just build it yourself using my LibreELEC fork (actually my last commit which includes the HyperHDR add-on). Follow the general instructions from LibreELEC, e.g. to build an add-on for Raspberry Pi 4 you need these two commands:

LibreELEC 11:
```
PROJECT=RPi ARCH=arm DEVICE=RPi4 make image
PROJECT=RPi DEVICE=RPi4 ARCH=arm ./scripts/create_addon hyperhdr
```

LibreELEC 12:
```
PROJECT=RPi ARCH=aarch64 DEVICE=RPi4 make image
PROJECT=RPi DEVICE=RPi4 ARCH=aarch64 ./scripts/create_addon hyperhdr
```

Sometimes it takes several executions to successfully build a base image: first command with 'make image', probably LibreELEC building nuances, it doesn't happen every time. The building process is intense and can take several hours! Although Ubuntu 22.04 LTS is not guaranteed by LibreELEC as a build host OS, it worked for me.

----

![obraz](https://github.com/awawa-dev/HyperHDR/assets/69086569/805691a0-f2f4-4e04-8609-a0400625f480)

![obraz](https://github.com/awawa-dev/HyperHDR/assets/69086569/ccd53f95-ebed-45de-8ef7-19ae63a60227)

**Comment by sjl68** [comment]
I installed this tonight on a intel system running LibreELEC 11.03, It all seems really good so far. Thank you for making this so easy! Some nice touches in the software from V19 as well.

**Comment by bruno-dogancic** [comment]
@awawa-dev Wow! This looks awesome!

After reading [your blog post](https://www.hyperhdr.eu/2022/05/tv-box-mania-iv-part-a95x-f3-air.html) I have remembered I also have A95X F3 Air device laying around and wanted to reuse my Rpi 4b 8Gb for better purposes; I have installed LE on it and then decided to go for Armbian as suggested in [your 2023 guide](https://www.hyperhdr.eu/2023/02/ultimate-guide-on-how-to-build-led.html#chapter2a).

Went through so much trouble getting back the stock firmware in order to run Armbian, only to find out this post as well 😄 
With that said, I didn't notice HyperHDR in the addons of LE when first exploring this device 😕 

Is there any difference in HyperHDR performance when run on LibreELEC or Armbian? On both devices, I had my Ethernet running, which is good since I have a switch behind my TV anyway, but on Armbian (now), the WiFi doesn't work, and I also really liked the ability for quick addons and the graphical interface of LibreELEC. Should I switch back to LE?

Also, if you can advise me on setting up HyperionScreenCap with HyperHDR... I cannot seem to forward the captured screen to the HyperHDR server, which runs on 192.168.1.173. The PC runs on 192.168.1.153.

Below are the screenshots

<details>

![image](https://github.com/awawa-dev/HyperHDR/assets/33345676/fc0eb96a-a93d-46e7-8060-ba33c282fc59)

![image](https://github.com/awawa-dev/HyperHDR/assets/33345676/5cf6bed1-8732-4ce1-9597-fcb7d6303a3a)

</details>

**Reply by awawa-dev** [reply]
Hi
If you had a USB3.0 grabber, I would choose Armbian. You can always add a wifi dongle from Aliexpress for a few dollars, just make sure that the given chipset model has been supported by Kernel for a long time. If you choose RGB/UDP color synchronization described below instead of flatbuffers, it wouldn't even have to be a 5/6g dongle, only 2.4g.

> Also, if you can advise me on setting up HyperionScreenCap with HyperHDR... I cannot seem to forward the captured screen to the HyperHDR server, which runs on 192.168.1.173. The PC runs on 192.168.1.153.

Make sure 19400 port is valid. Maybe wifi isolation is enabled or firewall? Ping from PC to RPI works? What is in the HyperHDR logs when connecting?
You can also use a much faster method instead of HyperionScreenCap/flatbuffers: run HyperHDR with built-in DirectX11 software grabber on PC, configure the same LED geometry as on RPI (or export/import RPi configuration) and output LED colors instead of large images inside flatbuffers frame. For flabuffers and a typical 640x480 frame, you need to send ~921 600 bytes over the network using the quite expensive TCP protocol, which, in addition to delay, also poses a risk of full WiFi saturation. For an alternative solution and sending only colors for ~300LED, it is only 900 bytes in one cheap UDP frame. Recently, a similar topic came back and you can start with it (there are links and a screenshot there): https://github.com/awawa-dev/HyperHDR/issues/673

**Comment by bruno-dogancic** [comment]
Hey and thank you for your time and fast reply. 🙏 

I think I will stick to Armbian then. I have this grabber, it claims it is USB3.0

<details>

![image](https://github.com/awawa-dev/HyperHDR/assets/33345676/1e769cc5-da48-4a1a-a4f3-37f07ccd83f7)
</details>

I have taken it because it has passthrough, but the the MS2130 is on the way as well as VSP01201 as I have only realized when testing that my netflix doesn't work with this one 😅 Trying to save money is usually equvivalent to wasting time 😄 

> Make sure 19400 port is valid. Maybe wifi isolation is enabled or firewall? Ping from PC to RPI works? What is in the HyperHDR logs when connecting? You can also use a much faster method instead of HyperionScreenCap/flatbuffers: run HyperHDR with built-in DirectX11 software grabber on PC, configure the same LED geometry as on RPI (or export/import RPi configuration) and output LED colors instead of large images inside flatbuffers frame. For flabuffers and a typical 640x480 frame, you need to send ~921 600 bytes over the network using the quite expensive TCP protocol, which, in addition to delay, also poses a risk of full WiFi saturation. For an alternative solution and sending only colors for ~300LED, it is only 900 bytes in one cheap UDP frame. Recently, a similar topic came back and you can start with it (there are links and a screenshot there): #673

Ping from Amlogic box to PC and vice versa works fine. I don't think I will use the afformentioned software once you told me for this UDP framework. I will try to fiddle with it. Many thanks for this one, still have so much to learn about HyperHDR 😃 

One last question: I used WLED in the old setup with Hyperion and Android Hyperion grabber. So it is my first time using USB grabbers. Is it normal that the USB grabber doesn't show within HyperHDR when I inserted it for the first time? I have even tried rebooting the device multiple time, and still the grabber didn't show in HyperHDR. It did show in the Armbian with either usb-devices and lsusb. I went to test it in windows machine in OBS and it worked just fine. Then when fiddling with HyperHDR on windows (to tryout UDP frame) I got the exact same behaviour. the USB grabber just doesn't show in HyperHDR. The workaround for this was to have the USB grabber connected and then forcing the restart using systemctl in Armbian and restaring the app in windows. After the restart the device shows and works notmally.

Also if someone will stumble on this post: [Rullz Coffee USB Grabber with passthrough USB 3.0 Loop Coffee](https://www.aliexpress.com/item/4001294726740.html?spm=a2g0o.order_list.order_list_main.33.3b181802SnTpxs) doesn't support HDCP ([checked with EME HDCP checker](https://googlechrome.github.io/samples/hdcp-detection/), only 1.0 which is kinda useless) and only gives the following values for capturing:

<details>

![image](https://github.com/awawa-dev/HyperHDR/assets/33345676/34db715d-9f0f-498b-810b-1d13c2341291)

![image](https://github.com/awawa-dev/HyperHDR/assets/33345676/8475bcde-0e23-43ec-ba9f-ed01b21fb794)

![image](https://github.com/awawa-dev/HyperHDR/assets/33345676/4ab0ee90-4db2-4e42-8364-57c9d544bc15)

</details>

**Comment by awawa-dev** [comment]
Plugins have been updated to latest HyperHDR v20beta1

**Reply by Johnny1982fd** [reply]
Works this with Hyper HDR 20 final and with LE 12 ?

**Comment by rschneider1509** [comment]
I also wanna know if there is a way to install on LE12?
The add-on does not work for me. Or I am doing smthn wrong

**Reply by awawa-dev** [reply]
LE12 is beta. As it says the addon is for stable LE11 but you can try to compile it yourself (generic manual in the first post here, LE12 is using aarch64 architectur for rpi4/5 https://github.com/awawa-dev/HyperHDR/discussions/831 )

**Reply by Johnny1982fd** [reply]
I think im too stupid for that. 

What does this mean?
PROJECT=RPi ARCH=aarch64 DEVICE=RPi4 make image
PROJECT=RPi DEVICE=RPi4 ARCH=aarch64 ./scripts/create_addon hyperhdr

Login via SSH to my LE 12 and put these 2 commands?

**Reply by rschneider1509** [reply]
nope, you have to "convert" the "TEXT" written by the developer to an usable software. this takes some experience, much time and CPU power. wont work with a pi.
i tried this on a Linux server at work, took a lot of time and did not succeed. i guess we have to wait for awawa to do this for us when the stable v12 releases.

i was also looking for any non-addon method to install in LE, but did not succeed. Main Problem is probably the LE switch to aarch64

**Reply by Johnny1982fd** [reply]
Ok, what a pity :(

LE 12 ist out of beta status since yesterday.

**Reply by awawa-dev** [reply]
I tried yesterday to prepare an environmental but currently there is a quite BIG problem with compiling QT library for LibreELEC 12 aarch64 (changed from arm in LE11) that is necessary for HyperHDR. Although I solved glibc bug in vector libs but community version Qt 5.15 is not compatible with  libxkbcommon 1.6.0 used in LibreELEC 12 and I'm unable to compile Qt6.2. ETA in unknown. If anyone has an idea to compile QT under LibreELEC 12 aarch64, I will be happy to hear it.

**Reply by rschneider1509** [reply]
I'm not a programmer and know way to less to contribute something useful here :(

**Reply by awawa-dev** [reply]
I will continue to test it this weekend. Probably Qt will need to be patched.

**Comment by awawa-dev** [comment]
Good news. I've managed to fix QT problems and have released HyperHDR add-ons for LibreELEC 12.
https://github.com/awawa-dev/LibreELEC.tv/releases/tag/v20_add_LE12
It would be nice if you could confirm that they work for you and which platform you are using (e.g. Rpi4/5/x64)

**Reply by Johnny1982fd** [reply]
I will test that, if it works it smells to a donation :)

**Reply by rschneider1509** [reply]
Works! <3

Thx for the fast support.
I did not find out, how to donate - will check this after  I watched the movie!

**Reply by Johnny1982fd** [reply]
https://github.com/sponsors/awawa-dev

**Reply by Johnny1982fd** [reply]
I see ... only Credit card?
PayPal possible ?

**Reply by awawa-dev** [reply]
Github changed something some time ago, so I don't know which payment channels are currently supported. Of course, donations are very nice and provide motivation to further develop the project, but giving a star to the project is also completely fair and enough, which increases HyperHDR reach on Github and lets others know that you are using our application.

**Reply by Johnny1982fd** [reply]
Just give your Paypal :)

**Comment by rschneider1509** [comment]
Btw, I am on rpi 4 64bit - LE12 aarch64.
And it works

**Comment by Mibix** [comment]
I tried the generic x64 version for a UGOO AM6B+ with CoreElec and can't seem to access the web address.  I already have HyperHDR running on my HTPC is there plugin that would just screen grab and send the data?  I have gotten TV-LED DV with FEL to work but my Elgato HD60 X is pretty much all green so was trying a few alternatives for when I have videos in that format.

EDIT:  I see you actually made a HyperHDR for CoreElec and that did work for capture till I played something in Dolby Vision.  I think my only option is to use Piccap.
